### DISCUSSION TITLE
`child_process` "pipe"s what are they really at the OS level, can they be used outside of Node.JS?

### DISCUSSION THREAD

**Discussion by segevfiner** [nodejs/help#4991] [Q&A]
The Node.JS docs state:
> 'pipe': Create a pipe between the child process and the parent process. The parent end of the pipe is exposed to the parent as a property on the child_process object as [subprocess.stdio[fd]](https://nodejs.org/api/child_process.html#subprocessstdio). Pipes created for fds 0, 1, and 2 are also available as [subprocess.stdin](https://nodejs.org/api/child_process.html#subprocessstdin), [subprocess.stdout](https://nodejs.org/api/child_process.html#subprocessstdout) and [subprocess.stderr](https://nodejs.org/api/child_process.html#subprocessstderr), respectively. Currently, these are not actual Unix pipes and therefore the child process can not use them by their descriptor files, e.g. /dev/fd/2 or /dev/stdout.
&ndash; https://nodejs.org/docs/latest-v16.x/api/child_process.html#optionsstdio

So what are they really? At the OS level? I'm looking for a way for a `child_process` to have a file descriptor/pipe it can monitor for closure to know when its parent process has terminated. But since those are not real file descriptors, I have no idea how I can use them from a child process that isn't Node.JS, or if it is even possible ATM. e.g. Can a Python process read from a `stdio > 2` spawned from a Node.JS parent?

**Comment by bnoordhuis** [comment]
They're UNIX sockets. Specifically, one end of a [socketpair(2)](https://man7.org/linux/man-pages/man2/socketpair.2.html).

**Reply by Antonius-S** [reply]
But how it works on Windows?

**Reply by bnoordhuis** [reply]
Named pipes.

**Reply by Antonius-S** [reply]
Okay, and how to use it with non-Node processes? What name, what the protocol is?
